libraries/standard/mqtt/src/mqtt_lightweight.c

```diff
@@ -1,16 +1,259 @@
+#include <string.h>
+
 #include "mqtt_lightweight.h"
 
+#define MQTT_VERSION_3_1_1    ( ( uint8_t ) 4U )
+
+#define MQTT_PACKET_CONNECT_MAX_SIZE    ( 327700UL )
+
+#define MQTT_CONNECT_FLAG_CLEAN          ( 1 )
+#define MQTT_CONNECT_FLAG_WILL           ( 2 )
+#define MQTT_CONNECT_FLAG_WILL_QOS1      ( 3 )
+#define MQTT_CONNECT_FLAG_WILL_QOS2      ( 4 )
+#define MQTT_CONNECT_FLAG_WILL_RETAIN    ( 5 )
+#define MQTT_CONNECT_FLAG_PASSWORD       ( 6 )
+#define MQTT_CONNECT_FLAG_USERNAME       ( 7 )
+
+#define UINT8_SET_BIT( x, position )    ( ( x ) = ( uint8_t ) ( ( x ) | ( 0x01U << ( position ) ) ) )
+
+#define UINT16_HIGH_BYTE( x )    ( ( uint8_t ) ( ( x ) >> 8 ) )
+#define UINT16_LOW_BYTE( x )     ( ( uint8_t ) ( ( x ) & 0x00ffU ) )
+
+static size_t remainingLengthEncodedSize( size_t length )
+{
+    size_t encodedSize;
+
+    /* Determine how many bytes are needed to encode length.
+     * The values below are taken from the MQTT 3.1.1 spec. */
+
+    /* 1 byte is needed to encode lengths between 0 and 127. */
+    if( length < 128U )
+    {
+        encodedSize = 1U;
+    }
+    /* 2 bytes are needed to encode lengths between 128 and 16,383. */
+    else if( length < 16384U )
+    {
+        encodedSize = 2U;
+    }
+    /* 3 bytes are needed to encode lengths between 16,384 and 2,097,151. */
+    else if( length < 2097152U )
+    {
+        encodedSize = 3U;
+    }
+    /* 4 bytes are needed to encode lengths between 2,097,152 and 268,435,455. */
+    else
+    {
+        encodedSize = 4U;
+    }
+
+    return encodedSize;
+}
+
+static uint8_t * encodeRemainingLength( uint8_t * pDestination,
+                                        size_t length )
+{
+    uint8_t lengthByte;
+    uint8_t * pLengthEnd = pDestination;
+    size_t remainingLength = length;
+
+    /* This algorithm is copied from the MQTT v3.1.1 spec. */
+    do
+    {
+        lengthByte = ( uint8_t ) ( remainingLength % 128U );
+        remainingLength = remainingLength / 128U;
+
+        /* Set the high bit of this byte, indicating that there's more data. */
+        if( remainingLength > 0U )
+        {
+            UINT8_SET_BIT( lengthByte, 7 );
+        }
+
+        /* Output a single encoded byte. */
+        *pLengthEnd = lengthByte;
```

Comment:
Does the caller ensure that the memory is not overflown?

Reply:
Yes